### PR TITLE
persist: add some more useful metrics

### PR DESCRIPTION
**Pull request by danhhz**
This adds prometheus metric coverage for encoding and decoding of state
(might have helped diagnose the CI slowness we investigated while trying
to land #13160). It also adds coverage for CaS retries as a result of
expectation mismatch (ditto).

### Motivation

  * This PR adds a feature that has not yet been specified.

    [Write a brief specification for the feature, including justification
     for its inclusion in Materialize, as if you were writing the original
     feature specification.]

### Testing

- [x] This PR has adequate test coverage / QA involvement has been duly considered.

### Release notes

This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - N/A

### CODE REVIEW

File: src/persist-client/src/impl/metrics.rs [hunk 1]
```diff
@@ -37,6 +37,8 @@ pub struct Metrics {
     pub cmds: CmdsMetrics,
     /// Metrics for each retry loop.
     pub retries: RetriesMetrics,
+    /// Metrics for various encodings and decodings.
+    pub codec: CodecMetrics,
```
Comment (danhhz) [inline src/persist-client/src/impl/metrics.rs:41]: oh! I should make this a set of vec counters with state, key, val as the labels. also I should add trace batch encode/decode to this PR

File: src/persist-client/src/impl/machine.rs [hunk 2]
```diff
@@ -279,6 +283,7 @@ where
     >(
         &mut self,
         cmd: &CmdMetrics,
+        metrics: &Metrics,
```
Comment (pH14) [inline src/persist-client/src/impl/machine.rs:286]: curious - is there a reason this needs to be passed in as a param vs accessing the metrics via `self`?

File: src/persist-client/src/impl/metrics.rs [hunk 3]
```diff
@@ -199,6 +235,23 @@ impl MetricsVecs {
         }
     }
 
+    fn codecs_metrics(&self) -> CodecsMetrics {
+        CodecsMetrics {
+            state: self.codec_metrics("state"),
+            batch: self.codec_metrics("batch"),
+            // Maybe {key,val}_{encode,decode} too. Probably not time or diff.
```
[the remark this answers is not in the document]
Comment (danhhz) [reply src/persist-client/src/impl/metrics.rs:242]: any thoughts on whether we should do this? easy enough to do but I wasn't sure if we thought the overhead of the measurement would be worth it

[the remark this answers is not in the document]
Comment (pH14) [reply src/persist-client/src/impl/metrics.rs:242]: unless we're concerned with metrics overhead, as a user I'd always be in favor of more metrics 😁

**Review by pH14 — APPROVED** [review]
LGTM

File: src/persist-client/src/impl/machine.rs (hunk 2 above)
Comment (danhhz) [reply src/persist-client/src/impl/machine.rs]: 🤦 it's been a week, apparently. done!

File: src/persist-client/src/impl/metrics.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (danhhz) [reply src/persist-client/src/impl/metrics.rs]: wfm! we can always rip it out later. added a helper for actually doing the measurements to CodecMetrics

**Review by danhhz — COMMENTED** [review]
TFTRs!

**Pull request merged**